sparse/core.py

```diff
@@ -603,55 +611,346 @@ def elemwise_binary(self, func, other, *args, **kwargs):
+            i = np.argsort(self_reduced_linear)
+            self_reduced_linear = self_reduced_linear[i]
+            self_coords = self_coords[:, i]
+            self_data = self_data[i]
+        except IndexError:
+            matches_found = False
 
         # Convert other.coords to a record array
-        other_coords = np.rec.fromarrays(other.coords)
+        other_coords = other.coords
         other_data = other.data
+        try:
+            other_reduced_coords, other_reduced_shape = \
+                self._get_reduced_coords(other_coords, other_shape,
+                                         other_reduce_params)
+            other_reduced_linear = self._linear_loc(other_reduced_coords, other_reduced_shape)
+            i = np.argsort(other_reduced_linear)
+            other_reduced_linear = other_reduced_linear[i]
+            other_coords = other_coords[:, i]
+            other_data = other_data[i]
+            pass
+        except IndexError:
+            matches_found = False
 
         # Find matches between self.coords and other.coords
-        j = np.searchsorted(self_coords, other_coords)
-        if len(self_coords):
-            matched_other = (other_coords == self_coords[j % len(self_coords)])
+        if matches_found:
+            matched_self, matched_other = _match_coords(self_reduced_linear,
+                                                        other_reduced_linear)
         else:
-            matched_other = np.zeros(shape=(0,), dtype=bool)
-        matched_self = j[matched_other]
+            matched_self, matched_other = np.repeat(np.arange(len(self_data)), len(other_data)), \
+                                          np.tile(np.arange(len(other_data)), len(self_data))
 
         # Locate coordinates without a match
-        unmatched_other = ~matched_other
-        unmatched_self = np.ones(len(self_coords), dtype=bool)
-        unmatched_self[matched_self] = 0
+        unmatched_self = np.ones(self.nnz, dtype=np.bool)
+        unmatched_self[matched_self] = False
+        unmatched_other = np.ones(other.nnz, dtype=np.bool)
+        unmatched_other[matched_other] = False
+
+        # Start with an empty list. This may reduce computation in many cases.
+        data_list = []
+        coords_list = []
+
+        # Add the matched part.
+        # if func(_TEST_NONZERO, _TEST_NONZERO_2):
+        matched_coords = self._get_matching_coords(self_coords[:, matched_self],
+                                                   other_coords[:, matched_other],
+                                                   self_shape, other_shape)
+
+        data_list.append(func(self_data[matched_self],
+                              other_data[matched_other],
+                              *args, **kwargs))
+        coords_list.append(matched_coords)
+
+        self_func = func(self_data, 0, *args, **kwargs)
+
+        # Add unmatched parts as necessary.
+        if (self_func != 0).any():
```

Comment:
Used this logic (same as below) to check if broadcasting "matches" are necessary. Got rid of the magic numbers. This should be solid.